volatility3/framework/objects/utility.py

```diff
@@ -29,29 +29,115 @@ def bswap_64(value: int) -> int:
     if not isinstance(array, objects.Array):
         raise TypeError("Array_to_string takes an Array of char")
 
     if count is None:
         count = array.vol.count
 
-    return array.cast("string", max_length=count, errors=errors)
+    return address_to_string(
+        context=array._context,
+        layer_name=array.vol.layer_name,
+        address=array.vol.offset,
+        count=count,
+        errors=errors,
+        block_size=block_size,
+    )
 
 
-def pointer_to_string(pointer: "objects.Pointer", count: int, errors: str = "replace"):
-    """Takes a volatility Pointer to characters and returns a string."""
+def pointer_to_string(
+    pointer: "objects.Pointer",
+    count: int,
+    errors: str = "replace",
+    block_size=32,
+) -> str:
+    """Takes a Volatility 'Pointer' to characters and returns a Python string.
+
+    Args:
+        pointer: A `Pointer` object containing character elements.
+        count: Optional maximum number of characters to convert. If None, the function
+               processes the entire array.
+        errors: Specifies error handling behavior for decoding, defaulting to "replace".
+        block_size: Reading block size. Defaults to 32
+
+    Returns:
+        A decoded string representation of the data referenced by the pointer.
+    """
     if not isinstance(pointer, objects.Pointer):
         raise TypeError("pointer_to_string takes a Pointer")
 
     if count < 1:
         raise ValueError("pointer_to_string requires a positive count")
 
-    char = pointer.dereference()
-    return char.cast("string", max_length=count, errors=errors)
+    return address_to_string(
+        context=pointer._context,
+        layer_name=pointer.vol.layer_name,
+        address=pointer,
+        count=count,
+        errors=errors,
+        block_size=block_size,
+    )
+
+
+def address_to_string(
+    context: interfaces.context.ContextInterface,
+    layer_name: str,
+    address: int,
+    count: int,
+    errors: str = "replace",
+    block_size=32,
+) -> str:
+    """Reads a null-terminated string from a given specified memory address, processing
+       it in blocks for efficiency.
+
+    Args:
+        context: The context used to retrieve memory layers and symbol tables
+        layer_name: The name of the memory layer to read from
+        address: The address where the string is located in memory
+        count: The number of bytes to read
+        errors: The error handling scheme to use for encoding errors. Defaults to "replace"
+        block_size: Reading block size. Defaults to 32
+
+    Returns:
+        The decoded string extracted from memory.
+    """
+    if not isinstance(address, int):
+        raise TypeError("It takes an int")
+
+    if count < 1:
+        raise ValueError("It requires a positive count")
+
+    layer = context.layers[layer_name]
+    text = b""
+    while len(text) <= count:
+        current_block_size = min(count - len(text), block_size)
```

Comment:
This needs to be positive (which I guess includes not 0).  Probably should improve the error checking in the physical layer too, but it would/should throw an exception so might as well fix it here too...